src/services/ftp/backend.rs

```diff
@@ -0,0 +1,512 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::fmt::Formatter;
+use std::io::Cursor;
+use std::io::Error;
+use std::io::ErrorKind;
+use std::io::Result;
+use std::io::SeekFrom;
+use std::str;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use async_compat::Compat;
+use async_trait::async_trait;
+use futures::AsyncRead;
+use futures::AsyncReadExt;
+use futures::AsyncSeekExt;
+use log::info;
+use suppaftp::list::File;
+use suppaftp::native_tls::TlsConnector;
+use suppaftp::FtpStream;
+use time::OffsetDateTime;
+
+use super::dir_stream::DirStream;
+use super::dir_stream::ReadDir;
+use super::err::new_request_append_error;
+use super::err::new_request_connection_error;
+use super::err::new_request_list_error;
+use super::err::new_request_mkdir_error;
+use super::err::new_request_put_error;
+use super::err::new_request_quit_error;
+use super::err::new_request_remove_error;
+use super::err::new_request_retr_error;
+use super::err::new_request_root_eror;
+use super::err::new_request_secure_error;
+use super::err::new_request_sign_error;
+use super::err::parse_io_error;
+use crate::error::other;
+use crate::error::BackendError;
+use crate::io_util::unshared_reader;
+use crate::ops::OpCreate;
+use crate::ops::OpDelete;
+use crate::ops::OpList;
+use crate::ops::OpRead;
+use crate::ops::OpStat;
+use crate::ops::OpWrite;
+use crate::Accessor;
+use crate::BytesReader;
+use crate::DirStreamer;
+use crate::ObjectMetadata;
+use crate::ObjectMode;
+
+/// Builder for ftp backend.
+#[derive(Default)]
+pub struct Builder {
+    endpoint: Option<String>,
+    port: Option<String>,
+    root: Option<String>,
+    user: Option<String>,
+    password: Option<String>,
+    tls: Option<String>,
```

Comment:
Just use bool.
`#[derive(Default)]` will make it `false` by default.

Reply:
Do you have other considerations for this?

And for naming, I prefer to use `enable_secure`.

Reply:
@Xuanwo 

I was thinking that making `tls`(or `enable_secure`) variable to be `str` type could enable user who use `.env` file to directly call `enable_secure()` function without any conversion just like:
```rust
builder.enable_secure(&env::var("OPENDAL_FTP_TLS"));
```
Otherwise, user who uses `.env` as configuration file need to first convert `enable_secure` to `bool` type first and then call `enable_secure()`.

Reply:
> Otherwise, user who uses `.env` as configuration file need to first convert `enable_secure` to `bool` type first and then call `enable_secure()`.

OpenDAL intends to be `painless`. We should not make users bother about which value I need to put. For boolean type options, OpenDAL prefer to use `enable_secure()` instead of `enable_secure(bool)`.

So it's the user's job to decide whether to call `enable_secure()` or not:

```rust
if let Ok(enable_secure) = env::var("OPENDAL_FTP_XXX") && enable_secure == "on" {
    builder.enable_secure();
}
```